File: src/rendering/PathRenderer.ts
import * as THREE from 'three';
import { HexCell } from '../types';
import { HexCoordinates } from '../core/HexCoordinates';
import { HexMetrics } from '../core/HexMetrics';

/**
 * Renders pathfinding visualization:
 * - Reachable cells (highlighted hexes)
 * - Path preview (line from unit to destination)
 */
export class PathRenderer {
  private scene: THREE.Scene;

  // Reachable cells visualization
  private reachableMeshes: THREE.InstancedMesh | null = null;
  private maxReachableInstances = 500;

  // Path line visualization
  private pathLine: THREE.Line | null = null;
  private pathLineMaterial: THREE.LineBasicMaterial;

  // Destination marker
  private destinationMarker: THREE.Mesh | null = null;

  constructor(scene: THREE.Scene) {
    this.scene = scene;

    // Create path line material
    this.pathLineMaterial = new THREE.LineBasicMaterial({
      color: 0x00ff00,
      linewidth: 2,
      transparent: true,
      opacity: 0.8,
    });

    this.createReachableMesh();
    this.createDestinationMarker();
  }

  /**
   * Create instanced mesh for reachable cells.
   */
  private createReachableMesh(): void {
    // Create a flat hexagon for highlighting
    const shape = new THREE.Shape();
    const radius = HexMetrics.outerRadius * 0.9;
    for (let i = 0; i < 6; i++) {
      const angle = (Math.PI / 3) * i - Math.PI / 6;
      const x = Math.cos(angle) * radius;
      const y = Math.sin(angle) * radius;
      if (i === 0) {
        shape.moveTo(x, y);
      } else {
        shape.lineTo(x, y);
      }
    }
    shape.closePath();

    const geometry = new THREE.ShapeGeometry(shape);
    geometry.rotateX(-Math.PI / 2);

    const material = new THREE.MeshBasicMaterial({
      color: 0x00ff00,
      transparent: true,
      opacity: 0.3,
      side: THREE.DoubleSide,
      depthWrite: false,
    });

    this.reachableMeshes = new THREE.InstancedMesh(
      geometry,
      material,
      this.maxReachableInstances
    );
    this.reachableMeshes.count = 0;
    this.reachableMeshes.frustumCulled = false;
    this.reachableMeshes.renderOrder = 1; // Render after terrain
    this.scene.add(this.reachableMeshes);
  }

  /**
   * Create destination marker mesh.
   */
  private createDestinationMarker(): void {
    const geometry = new THREE.RingGeometry(0.6, 0.8, 6);
    geometry.rotateX(-Math.PI / 2);
    geometry.rotateY(Math.PI / 6);

    const material = new THREE.MeshBasicMaterial({
      color: 0xff0000,
      transparent: true,
      opacity: 0.8,
      side: THREE.DoubleSide,
      depthWrite: false,
    });

    this.destinationMarker = new THREE.Mesh(geometry, material);
    this.destinationMarker.visible = false;
    this.destinationMarker.renderOrder = 2;
    this.scene.add(this.destinationMarker);
  }

  /**
   * Show reachable cells for a unit.
   */
  showReachableCells(reachableCells: Map<HexCell, number>): void {
    if (!this.reachableMeshes) return;

    const matrix = new THREE.Matrix4();
    const color = new THREE.Color();
    let index = 0;

    for (const [cell, cost] of reachableCells) {
      if (index >= this.maxReachableInstances) break;

      const coords = new HexCoordinates(cell.q, cell.r);
      // For water cells (elevation < 0), render at water surface level
      const renderElevation = Math.max(cell.elevation, 0);
      const worldPos = coords.toWorldPosition(renderElevation);

      matrix.makeTranslation(worldPos.x, worldPos.y + 0.05, worldPos.z);
      this.reachableMeshes.setMatrixAt(index, matrix);

      // Color based on movement cost (green = cheap, yellow = expensive)
      const t = Math.min(cost / 4, 1); // Normalize to 0-1 over 4 movement points
      color.setHSL(0.33 - t * 0.33, 0.8, 0.5); // Green to yellow
      this.reachableMeshes.setColorAt(index, color);

      index++;
    }

    this.reachableMeshes.count = index;
    this.reachableMeshes.instanceMatrix.needsUpdate = true;
    if (this.reachableMeshes.instanceColor) {
      this.reachableMeshes.instanceColor.needsUpdate = true;
    }
  }

  /**
   * Hide reachable cells.
   */
  hideReachableCells(): void {
    if (this.reachableMeshes) {
      this.reachableMeshes.count = 0;
    }
  }

  /**
   * Show path preview from unit to destination.
   */
  showPath(path: HexCell[]): void {
    // Remove old path line
    if (this.pathLine) {
      this.scene.remove(this.pathLine);
      this.pathLine.geometry.dispose();
      this.pathLine = null;
    }

    if (path.length < 2) {
      this.hideDestinationMarker();
      return;
    }

    // Create path points
    const points: THREE.Vector3[] = [];
    for (const cell of path) {
      const coords = new HexCoordinates(cell.q, cell.r);
      // For water cells (elevation < 0), render at water surface level
      const renderElevation = Math.max(cell.elevation, 0);
      const worldPos = coords.toWorldPosition(renderElevation);
      points.push(new THREE.Vector3(worldPos.x, worldPos.y + 0.2, worldPos.z));
    }

    // Create line geometry
    const geometry = new THREE.BufferGeometry().setFromPoints(points);
    this.pathLine = new THREE.Line(geometry, this.pathLineMaterial);
    this.pathLine.renderOrder = 2;
    this.scene.add(this.pathLine);

    // Show destination marker at end of path
    const lastCell = path[path.length - 1];
    this.showDestinationMarker(lastCell);
  }

  /**
   * Hide path preview.
   */
  hidePath(): void {
    if (this.pathLine) {
      this.scene.remove(this.pathLine);
      this.pathLine.geometry.dispose();
      this.pathLine = null;
    }
    this.hideDestinationMarker();
  }

  /**
   * Show destination marker at a cell.
   */
  private showDestinationMarker(cell: HexCell): void {
    if (!this.destinationMarker) return;

    const coords = new HexCoordinates(cell.q, cell.r);
    // For water cells (elevation < 0), render at water surface level
    const renderElevation = Math.max(cell.elevation, 0);
    const worldPos = coords.toWorldPosition(renderElevation);
    this.destinationMarker.position.set(worldPos.x, worldPos.y + 0.1, worldPos.z);
    this.destinationMarker.visible = true;
  }

  /**
   * Hide destination marker.
   */
  private hideDestinationMarker(): void {
    if (this.destinationMarker) {
      this.destinationMarker.visible = false;
    }
  }

  /**
   * Update path line color (for valid/invalid paths).
   */
  setPathValid(valid: boolean): void {
    this.pathLineMaterial.color.set(valid ? 0x00ff00 : 0xff0000);
  }

  /**
   * Clean up resources.
   */
  dispose(): void {
    if (this.reachableMeshes) {
      this.scene.remove(this.reachableMeshes);
      this.reachableMeshes.geometry.dispose();
      (this.reachableMeshes.material as THREE.Material).dispose();
      this.reachableMeshes = null;
    }

    if (this.pathLine) {
      this.scene.remove(this.pathLine);
      this.pathLine.geometry.dispose();
      this.pathLine = null;
    }

    this.pathLineMaterial.dispose();

    if (this.destinationMarker) {
      this.scene.remove(this.destinationMarker);
      this.destinationMarker.geometry.dispose();
      (this.destinationMarker.material as THREE.Material).dispose();
      this.destinationMarker = null;
    }
  }
}

File: results/csharp_converted/organized/Pathfinding/path_renderer.cs
using Godot;
using Godot.Collections;


//# Renders pathfinding visualization:
//# - Reachable cells (highlighted hexes)
//# - Path preview (line from unit to destination)

//# Matches web/src/rendering/PathRenderer.ts
[GlobalClass]
public partial class PathRenderer : Godot.Node3D
{
	public Godot.HexGrid Grid;


	// Reachable cells visualization
	public Godot.MultiMeshInstance3D ReachableMeshes;
	public int MaxReachableInstances = 500;


	// Path line visualization
	public Godot.MeshInstance3D PathLine;
	public Godot.StandardMaterial3D PathMaterial;


	// Destination marker
	public Godot.MeshInstance3D DestinationMarker;


	public override void _Init()
	{
		_CreateReachableMesh();
		_CreatePathMaterial();
		_CreateDestinationMarker();
	}


	public void Setup(Godot.HexGrid p_grid)
	{
		System.Diagnostics.Debug.Assert(p_grid != null, "PathRenderer requires HexGrid");

		Grid = p_grid;
	}


	protected void _CreateReachableMesh()
	{

		// Create a flat hexagon for highlighting
		var mesh = _BuildHexShapeMesh();

		var multimesh = MultiMesh.New();
		multimesh.TransformFormat = MultiMesh.TransformFormat.Transform3d;
		multimesh.UseColors = true;
		multimesh.Mesh = mesh;
		multimesh.InstanceCount = MaxReachableInstances;
		multimesh.VisibleInstanceCount = 0;

		ReachableMeshes = MultiMeshInstance3D.New();
		ReachableMeshes.Multimesh = multimesh;

		var material = StandardMaterial3D.New();
		material.AlbedoColor = new Color(0.0, 1.0, 0.0, 0.5);
		material.Transparency = BaseMaterial3D.Transparency.TransparencyAlpha;
		material.ShadingMode = BaseMaterial3D.ShadingMode.ShadingModeUnshaded;
		material.CullMode = BaseMaterial3D.CullMode.CullDisabled;
		material.NoDepthTest = true;
		// Render on top of terrain
		material.VertexColorUseAsAlbedo = true;
		ReachableMeshes.MaterialOverride = material;

		AddChild(ReachableMeshes);
	}


	protected Godot.ArrayMesh _BuildHexShapeMesh()
	{
		var st = SurfaceTool.New();
		st.Begin(Mesh.PrimitiveType.PrimitiveTriangles);

		var radius = HexMetrics.OUTER_RADIUS * 0.9;
		var corners = new Array{};

		foreach(int i in GD.Range(6))
		{
			var angle = (Mathf.Pi / 3.0) * i - Mathf.Pi / 6.0;
			corners.Append(new Vector3(Mathf.Cos(angle) * radius, 0, Mathf.Sin(angle) * radius));
		}


		// Build triangles from center
		var center = Vector3.Zero;
		st.SetNormal(Vector3.Up);
		foreach(int i in GD.Range(6))
		{
			st.AddVertex(center);
			st.AddVertex(corners[i]);
			st.AddVertex(corners[(i + 1) % 6]);
		}

		return st.Commit();
	}


	protected void _CreatePathMaterial()
	{
		PathMaterial = StandardMaterial3D.New();
		PathMaterial.AlbedoColor = new Color(0.0, 1.0, 0.0, 0.8);
		PathMaterial.Transparency = BaseMaterial3D.Transparency.TransparencyAlpha;
		PathMaterial.ShadingMode = BaseMaterial3D.ShadingMode.ShadingModeUnshaded;
	}


	protected void _CreateDestinationMarker()
	{

		// Create a ring mesh for destination
		var st = SurfaceTool.New();
		st.Begin(Mesh.PrimitiveType.PrimitiveTriangles);

		var inner_radius = 0.6;
		var outer_radius = 0.8;

		foreach(int i in GD.Range(6))
		{
			var angle1 = (Mathf.Pi / 3.0) * i - Mathf.Pi / 6.0;
			var angle2 = (Mathf.Pi / 3.0) * ((i + 1) % 6) - Mathf.Pi / 6.0;

			var inner1 = new Vector3(Mathf.Cos(angle1) * inner_radius, 0, Mathf.Sin(angle1) * inner_radius);
			var outer1 = new Vector3(Mathf.Cos(angle1) * outer_radius, 0, Mathf.Sin(angle1) * outer_radius);
			var inner2 = new Vector3(Mathf.Cos(angle2) * inner_radius, 0, Mathf.Sin(angle2) * inner_radius);
			var outer2 = new Vector3(Mathf.Cos(angle2) * outer_radius, 0, Mathf.Sin(angle2) * outer_radius);

			st.SetNormal(Vector3.Up);
			st.AddVertex(inner1);
			st.AddVertex(outer1);
			st.AddVertex(outer2);

			st.AddVertex(inner1);
			st.AddVertex(outer2);
			st.AddVertex(inner2);
		}

		DestinationMarker = MeshInstance3D.New();
		DestinationMarker.Mesh = st.Commit();

		var material = StandardMaterial3D.New();
		material.AlbedoColor = new Color(1.0, 0.0, 0.0, 0.8);
		material.Transparency = BaseMaterial3D.Transparency.TransparencyAlpha;
		material.ShadingMode = BaseMaterial3D.ShadingMode.ShadingModeUnshaded;
		material.CullMode = BaseMaterial3D.CullMode.CullDisabled;
		DestinationMarker.MaterialOverride = material;
		DestinationMarker.Visible = false;

		AddChild(DestinationMarker);
	}


	//# Show reachable cells for a unit
	public void ShowReachableCells(Dictionary reachable_cells)
	{
		if(ReachableMeshes == null)
		{
			return ;
		}

		var mm = ReachableMeshes.Multimesh;
		var index = 0;

		foreach(Variant cell in reachable_cells.Keys())
		{
			if(index >= MaxReachableInstances)
			{
				break;
			}

			var cost = reachable_cells[cell];
			var world_pos = HexCoordinates.New(cell.Q, cell.R).ToWorldPosition(0);


			// For water cells, render on water surface; for land, render on terrain
			var y_offset;
			if(cell.Elevation < HexMetrics.SEA_LEVEL)
			{
				y_offset = HexMetrics.SEA_LEVEL * HexMetrics.ELEVATION_STEP + 0.1;
			}
			// Just above water surface
			else
			{
				y_offset = cell.Elevation * HexMetrics.ELEVATION_STEP + 0.15;
			}

			var transform = new Transform3D();
			Transform.Origin = new Vector3(world_pos.X, y_offset, world_pos.Z);
			mm.SetInstanceTransform(index, Transform);


			// Color based on movement cost (green = cheap, yellow = expensive)
			var t = Mathf.Min(cost / 4.0, 1.0);
			var color = Color.FromHsv(0.33 - t * 0.33, 0.8, 0.7, 0.5);
			mm.SetInstanceColor(index, color);

			index += 1;
		}

		mm.VisibleInstanceCount = index;
	}


	//# Hide reachable cells
	public void HideReachableCells()
	{
		if(ReachableMeshes && ReachableMeshes.Multimesh)
		{
			ReachableMeshes.Multimesh.VisibleInstanceCount = 0;
		}
	}


	//# Show path preview from unit to destination
	public void ShowPath(Array path)
	{

		// Remove old path line
		if(PathLine)
		{
			PathLine.QueueFree();
			PathLine = null;
		}

		if(path.Size() < 2)
		{
			HideDestinationMarker();
			return ;
		}


		// Create path points
		var points = new Array{};
		foreach(Variant cell in path)
		{
			var world_pos = HexCoordinates.New(cell.Q, cell.R).ToWorldPosition(0);
			var y_pos;
			if(cell.Elevation < HexMetrics.SEA_LEVEL)
			{
				y_pos = HexMetrics.SEA_LEVEL * HexMetrics.ELEVATION_STEP + 0.15;
			}
			// Above water surface
			else
			{
				y_pos = cell.Elevation * HexMetrics.ELEVATION_STEP + 0.2;
			}
			points.Append(new Vector3(world_pos.X, y_pos, world_pos.Z));
		}


		// Create line mesh using ImmediateMesh
		var im = ImmediateMesh.New();
		im.SurfaceBegin(Mesh.PrimitiveType.PrimitiveLineStrip);
		foreach(PackedVector3Array point in points)
		{
			im.SurfaceAddVertex(point);
		}
		im.SurfaceEnd();

		PathLine = MeshInstance3D.New();
		PathLine.Mesh = im;
		PathLine.MaterialOverride = PathMaterial;
		AddChild(PathLine);


		// Show destination marker at end of path
		var last_cell = path[path.Size() - 1];
		ShowDestinationMarker(last_cell);
	}


	//# Hide path preview
	public void HidePath()
	{
		if(PathLine)
		{
			PathLine.QueueFree();
			PathLine = null;
		}
		HideDestinationMarker();
	}


	//# Show destination marker at a cell
	public void ShowDestinationMarker(Godot.HexCell cell)
	{
		if(DestinationMarker == null)
		{
			return ;
		}

		var world_pos = HexCoordinates.New(cell.Q, cell.R).ToWorldPosition(0);
		var y_pos;
		if(cell.Elevation < HexMetrics.SEA_LEVEL)
		{
			y_pos = HexMetrics.SEA_LEVEL * HexMetrics.ELEVATION_STEP + 0.12;
		}
		// Above water surface
		else
		{
			y_pos = cell.Elevation * HexMetrics.ELEVATION_STEP + 0.15;
		}
		DestinationMarker.Position = new Vector3(world_pos.X, y_pos, world_pos.Z);
		DestinationMarker.Visible = true;
	}


	//# Hide destination marker
	public void HideDestinationMarker()
	{
		if(DestinationMarker)
		{
			DestinationMarker.Visible = false;
		}
	}


	//# Update path line color (for valid/invalid paths)
	public void SetPathValid(bool valid)
	{
		if(PathMaterial)
		{
			PathMaterial.AlbedoColor = ( valid ? new Color(0.0, 1.0, 0.0, 0.8) : new Color(1.0, 0.0, 0.0, 0.8) );
		}
	}


}